### ISSUE TITLE
Does bonnetal classification module support multilabel classification?

### ISSUE THREAD

**Issue by newforrestgump001** [PRBonn/bonnetal#7]
Does bonnetal classification module support multilabel classification?
Thanks a lot!
Jack

**Comment by tano297** [comment]
Hi,
As it is, it doesn't. BUT, it should be straightforward to implement.
Changes that would be necessary:

- During training: changing Softmax and CrossEntropy with Sigmoid and BinaryCrossEntropy
- During inference: changing Softmax for Sigmoid, and replacing the Argmax operation (which gets only the highest class), for a threshold in the confidence of your choice, and then a selection of all the classes that correspond with this criteria.

If you're interested in making this happen neatly, I can give you pointers, and maybe you can submit a PR.

Best,

**Comment by newforrestgump001** [comment]
Thanks a lot for your information.
My filelist.txt is like this:
4590.jpg 1 1 0 0
7331.jpg 1 0 1 0
4941.jpg 0 1 1 0
1030.jpg 0 1 1 0
3094.jpg 1 1 1 0
8528.jpg 1 0 0 0
1314.jpg 1 0 0 0
5335.jpg 0 1 1 0
4097.jpg 0 0 0 1
7059.jpg 1 0 0 1
2858.jpg 1 1 1 1
2972.jpg 0 0 1 0
2070.jpg 0 0 1 0
3927.jpg 1 0 0 1
3106.jpg 1 0 0 0
6214.jpg 1 1 0 0
5578.jpg 0 1 0 0
3766.jpg 1 1 1 1
6954.jpg 1 1 1 1
6997.jpg 1 1 1 1
8901.jpg 0 1 0 0
632.jpg 1 0 1 0
1799.jpg 0 0 1 0
5157.jpg 1 0 1 0
2456.jpg 0 1 1 0
7542.jpg 1 1 1 1
9093.jpg 0 1 1 0
2956.jpg 0 1 0 0
680.jpg 1 0 1 0
1441.jpg 1 1 1 1
Could you give me some pointers about how to modify the framework  on the main points? Thanks a lot!

**Comment by newforrestgump001** [comment]
@tano297 Could you kindly give me some pointers for the modification of the classification module to support multi-label classification, I'm new to bonnetal framework. Thanks a lot!

**Comment by tano297** [comment]
- [criterion](https://github.com/PRBonn/bonnetal/blob/5368eed73935712af161e3d93680ccab0eeb64ec/train/tasks/classification/modules/trainer.py#L125) needs to be modified to allow for multiple labels. See pytorch [BCEWithLogitsLoss](https://pytorch.org/docs/stable/nn.html#torch.nn.BCEWithLogitsLoss)
- [user.py tok's](https://github.com/PRBonn/bonnetal/blob/5368eed73935712af161e3d93680ccab0eeb64ec/train/tasks/classification/modules/user.py#L145) needs to be replaced with a threshold and a torch.nonzero in order to get all the positive labels.
- Your parser needs to output the labels in your multilabel format, rather than as a single number, in order to make the BCEwithlogitsloss work properly

**Comment by newforrestgump001** [comment]
@tano297 Thanks for your help, I will have a try.
Jack

**Comment by tano297** [comment]
I'll close this for now, let me know if you need more pointers

**Comment by newforrestgump001** [comment]
@tano297 Thanks, you're very nice!
Jack
